agent/checkpoint.go

```diff
@@ -0,0 +1,33 @@
+package agent
+
+import (
+	"context"
+)
+
+// checkpoint allows a goroutine to communicate when it is OK to proceed beyond some async condition
+// to other dependent goroutines.
+type checkpoint struct {
+	done chan struct{}
+	err  error
+}
+
+// complete the checkpoint.  Pass nil to indicate the checkpoint was ok.
+func (c *checkpoint) complete(err error) {
+	c.err = err
```

Comment:
How about using sync once here to simplify usage an allowing multiple calls to complete?

Reply:
That introduces the possibility of multiple calls racing success vs failure.  Better to ensure that there is only one call to `complete()`.

Reply:
In what way would it be racy? Sync once is atomic AFAIK and first come-first served, the other remains blocked waiting for the first to complete.

Reply:
Perhaps another way to put it: I think it would be preferable for this to be safe to use incorrectly and document how it _should_ be used, vs incorrect use resulting in a runtime panic.

Reply:
It's racy in the sense that if you have two callers, one saying "success" and one saying "failure", then they can race each other to determine which gets reported to things waiting on the checkpoint.

Reply:
I think I understand your concern but I have a hard time understanding how it's relevant here. If someone introduces that race right now the program will panic instead. So it needs to be avoided in either case, i.e. a non-issue.

Reply:
> I think it would be preferable for this to be safe to use incorrectly and document how it should be used, vs incorrect use (or carelessness) resulting in a runtime panic.

I think that kind of defensive programming is a disservice in this case.  Silently absorbing incorrect/careless use in a non-deterministic way is a source of frustrating and hard to diagnose bugs.

There is a genuine case to be made for a central service like Coderd needing to avoid panicking because of programming bugs, but here in the Agent, I think it's preferable to panic, print a stack trace, and exit.

Reply:
I can definitely see the benefit of that approach as well, but I do feel we should have a high tolerance towards introducing sharp edges that can result in decreased developer flow. If we feel this is important enough to panic, wouldn’t another approach be to detect a second call and log it as an error + stack trace?

We don’t know how well the logs from a panic will be persisted either, the user may attempt to rebuild their workspace and ultimately we caused an inconvenience and are never the wiser.